index.js

```diff
@@ -300,10 +301,76 @@ const vfsActionFactory = (core, proc, win, dialog, state) => {
     });
   };
 
-  const uploadBrowserFiles = (files) => {
-    Promise.all(files.map(writeRelative))
-      .then(() => refresh(files[0].name)) // FIXME: Select all ?
-      .catch(error => dialog('error', error, __('MSG_UPLOAD_ERROR')));
+  const uploadBrowserFiles = async items => {
+    const uploadList = [];
+    let totalSize = 0;
+    let filename;
+    try {
+      const checkFile = (file, dirPath) => {
+        uploadList.push({dirPath, file});
+        totalSize += file.size;
+        if (filename === undefined) {
+          filename = file.name;
+        } else if (filename) {
+          filename = '';
+        }
+      };
+      const checkDirectory = async (directory, dirPath) => {
+        uploadList.push({dirPath});
+        const reader = directory.createReader();
+        await new Promise(resolve => {
+          reader.readEntries(async entries => {
+            for (let entry of entries) {
+              if (entry.isFile) {
+                await new Promise((resolve, reject) => {
+                  entry.file(file => {
+                    checkFile(file, dirPath);
+                    resolve();
+                  }, reject);
+                });
+              } else if (entry.isDirectory) {
+                await checkDirectory(entry, dirPath + '/' + entry.name);
+              }
+            }
+            resolve();
+          });
+        });
+      };
+      for (let item of items) {
+        const entry = item.webkitGetAsEntry();
```

Comment:
Just had a quick look through these changes and this method call caught my attention because of the prefix.

Doesn't seem like this works on Firefox on Android 🤔 

Ref: https://developer.mozilla.org/en-US/docs/Web/API/DataTransferItem/webkitGetAsEntry#browser_compatibility

Reply:
Yes. It's not exactly standard and supported everywhere 😓. Perhaps we could check for support first and if not, revert to the old way?

Reply:
Indeed!

Reply:
done